synapse/handlers/room_member.py

```diff
@@ -498,35 +500,63 @@ def _update_membership(
         return res
 
     @defer.inlineCallbacks
-    def copy_user_state_if_room_upgrade(self, new_room_id, user_id):
-        """Copy user-specific information when they join a new room if that new room is the
-        result of a room upgrade
+    def transfer_room_state_if_room_upgrade(self, room_id, user_id):
+        """Upon a user joining a room, the server can check if that room is the result of a
+        room upgrade. At this point, the server can transfer over information from the previous
+        room.
+
+        Checks if the given room is the result of a room upgrade. If so, copies user state
+        (tags/push rules) from the old room, as well as depublishing it from the room
+        directory if it was in there.
 
         Args:
-            new_room_id (str): The ID of the room the user is joining
+            room_id (str): The ID of the room the user is joining
+
             user_id (str): The ID of the user
 
         Returns:
             Deferred
         """
         # Check if the new room is an upgraded room
-        predecessor = yield self.store.get_room_predecessor(new_room_id)
+        predecessor = yield self.store.get_room_predecessor(room_id)
         if not predecessor:
             return
+        old_room_id = predecessor["room_id"]
+        logger.debug("Found predecessor for %s: %s", room_id, old_room_id)
+
+        # Copy over user state if this is a join on an remote upgraded room
+        yield self.copy_user_state_on_room_upgrade(old_room_id, room_id, user_id)
+
+        # Depublish the old room from the room directory if it was published
```

Comment:
It doesn't, it was just the most convenient place to put it. It could probably happen after the room join as well.

Is this to in referencing to keeping room joins quick?

Reply:
Mostly, I'd just expect it to happen as part of tombstoning the old room. It just seems odd for this to happen every time a user joins the new room.

Reply:
> I guess so. Another option might be to do it when you process the create for the new room?

Sure, though what would be the advantage over doing so on join then? Speed of join still?

Reply:
Ah, yeah good point. I'll see if I can find a suitable hook.

Reply:
Hooks have been changed to on joining a room over federation, and on upgrading the room ourselves.